### PR TITLE
[CELEBORN-1073] Avoid memory copying in TransportMessage.toByteBuffer by using wrapping payload into CompositeByteBuf

### PR DESCRIPTION
**Pull request by bowenliang123**
### What changes were proposed in this pull request?
TransportMessage.toByteBuffer copies its payload byte array to a new dedicated byte array. Change to use Netty's CompositeByteBuf helps to reuse the payload array and skip redundant byte array copying and for less memory footprint. And also prevents manually shifting offset (+4 +4) for integers of the message header.

### Why are the changes needed?
To skip redundant byte array copying and for less memory footprint.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

CI tests.

### CODE REVIEW

File: common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java [hunk 1]
```diff
@@ -97,13 +99,11 @@ public <T extends GeneratedMessageV3> T getParsedPayload() throws InvalidProtoco
...
-    ByteBuffer buffer = ByteBuffer.allocate(totalBufferSize);
-    buffer.putInt(messageTypeValue);
-    buffer.putInt(payload.length);
-    buffer.put(payload);
-    buffer.flip();
-    return buffer;
+    CompositeByteBuf buffer = Unpooled.compositeBuffer();
+    buffer.writeInt(messageTypeValue);
+    buffer.writeInt(payload.length);
+    buffer.writeBytes(Unpooled.wrappedBuffer(payload));
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java:105]: image
Hi @bowenliang123 , I debugged into this code and find out that although `buffer.nioBuffer()` avoids memory copy, `buffer.writeBytes(Unpooled.wrappedBuffer(payload));` involves memory copy. And the logic becomes more complex.

Comment (bowenliang123) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java:105]: Hi, thanks for the feedback. May I ask which class is this screenshot for?

Comment (bowenliang123) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java:105]: We should change to use `addComponent(boolean, ByteBuf)` to append the wrapped buffer. But somehow it failed the CI tests.

Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java:105]: I ran `HugeDataTest` and the snapshot is from `UnpooledHeapByteBuf#setBytes`

Comment (bowenliang123) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java:105]: image

Comment (bowenliang123) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java]: It seems that with the `CompositeByteBuf` (or `wrappedBuffer()`) , getting NIO buffer by calling the `nioBuffer` method always cause new byte buffer allocation and byte copying.

Comment (bowenliang123) [reply common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java]: If we keep the current code changes, while updating the PR purpose for avoid manually controlling offset in `TransparentMessasge`'s NIO buffer assembling, would you like to consider this PR? @waitinfuture

File: common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java [hunk 2]
```diff
@@ -97,13 +99,10 @@ public <T extends GeneratedMessageV3> T getParsedPayload() throws InvalidProtoco
...
 
   public ByteBuffer toByteBuffer() {
-    int totalBufferSize = payload.length + 4 + 4;
-    ByteBuffer buffer = ByteBuffer.allocate(totalBufferSize);
-    buffer.putInt(messageTypeValue);
-    buffer.putInt(payload.length);
-    buffer.put(payload);
-    buffer.flip();
-    return buffer;
+    ByteBuf headerBuf = Unpooled.buffer().writeInt(messageTypeValue).writeInt(payload.length);
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java:102]: `Unpooled.buffer()` allocates 256 bytes as default `static final int DEFAULT_INITIAL_CAPACITY = 256;`, which is much larger than the needed 8 bytes. Additionally, `toByteBuffer` is used for few messages whose payload is not big, IMO it's OK to leave what it is.

**Pull request closed without merging**